chips/stm32f303xc/src/flash.rs

```diff
@@ -0,0 +1,636 @@
+                FlashState::Erase => {
+                    self.client.map(|client| {
+                        client.erase_complete(hil::flash::Error::FlashError);
+                    });
+                }
+                _ => {}
+            }
+
+            self.state.set(FlashState::Ready);
+        }
+    }
+
+    pub fn program_halfword(&self) {
+        self.buffer.take().map(|buffer| {
+            let i = self.write_counter.get();
+
+            let halfword: u16 = (buffer[i] as u16) << 0 | (buffer[i + 1] as u16) << 8;
+            let page_addr = PAGE_START + self.page_number.get() * PAGE_SIZE;
+            let address = page_addr + i;
+            let location = unsafe { &*(address as *const VolatileCell<u16>) };
+            location.set(halfword);
+
+            self.buffer.replace(buffer);
+        });
+    }
+
+    pub fn erase_page(&self, page_number: usize) -> ReturnCode {
+        if page_number > 128 {
+            return ReturnCode::EINVAL;
+        }
+
+        if self.is_locked() {
+            self.unlock();
+        }
+
+        self.enable();
+        self.state.set(FlashState::Erase);
+
+        // Choose page erase mode.
+        self.registers.cr.modify(Control::PER::SET);
+        self.registers
+            .ar
+            .write(Address::FAR.val((PAGE_START + page_number * PAGE_SIZE) as u32));
+        self.registers.cr.modify(Control::STRT::SET);
+
+        ReturnCode::SUCCESS
+    }
+
+    pub fn erase_all(&self) -> ReturnCode {
+        if self.is_locked() {
+            self.unlock();
+        }
+
+        self.enable();
+        self.state.set(FlashState::Erase);
+
+        // Choose mass erase mode.
+        self.registers.cr.modify(Control::MER::SET);
+        self.registers.cr.modify(Control::STRT::SET);
+
+        ReturnCode::SUCCESS
+    }
+
+    pub fn write_page(
+        &self,
+        page_number: usize,
+        buffer: &'static mut StmF303Page,
+    ) -> Result<(), (ReturnCode, &'static mut StmF303Page)> {
+        if page_number > 128 {
+            return Err((ReturnCode::EINVAL, buffer));
+        }
+
+        if self.is_locked() {
+            self.unlock();
+        }
+
+        self.enable();
+        self.state.set(FlashState::Write);
+
+        // Choose programming mode.
+        self.registers.cr.modify(Control::PG::SET);
+
+        self.buffer.replace(buffer);
+        self.page_number.set(page_number);
+        self.program_halfword();
+
+        Ok(())
+    }
+
+    pub fn read_page(
+        &self,
+        page_number: usize,
+        buffer: &'static mut StmF303Page,
+    ) -> Result<(), (ReturnCode, &'static mut StmF303Page)> {
+        if page_number > 128 {
+            return Err((ReturnCode::EINVAL, buffer));
+        }
+
+        let mut byte: *const u8 = (PAGE_START + page_number * PAGE_SIZE) as *const u8;
+        unsafe {
+            for i in 0..buffer.len() {
+                buffer[i] = *byte;
+                byte = byte.offset(1);
+            }
+        }
```

Comment:
I have just realised that I am only doing normal pointer reads here.  The reference manual states the following "The embedded Flash module can be addressed directly, as a common memory space.", which makes me think that it is fine as it is. Do you think I should make them volatile?

Reply:
I think they only need to be volatile if there is a chance that the memory values will change from something _other_ than the executing code. That's necessary when doing DMA with peripherals, where a sensor/radio/etc might independently write to memory. As the flash controller won't change memory, I don't think volatile is required here.

Reply:
Makes sense.